compiler/passes/ResolveScope.cpp

```diff
@@ -451,6 +451,147 @@ bool ResolveScope::isSymbolAndMethod(Symbol* sym0, Symbol* sym1) {
+************************************** | *************************************/
+
+Symbol* ResolveScope::lookupNameLocallyForImport(const char* name) const {
+  Bindings::const_iterator it     = mBindings.find(name);
+  Symbol*                  retval = NULL;
+
+  if (it != mBindings.end()) {
+    retval = it->second;
+  }
+  return retval;
+}
+
+Symbol* ResolveScope::followImportUseChains(UnresolvedSymExpr* expr) const {
+  const char* name = expr->unresolved;
+  UseImportList useImportList = mUseImportList;
+  SymList symbols;
+
+  buildBreadthFirstUseImportList(useImportList);
+
+  // Do not use for_vector(); it terminates on a NULL
+  for (size_t i = 0; i < useImportList.size(); i++) {
+    if (UseStmt* use = toUseStmt(useImportList[i])) {
+      if (use->skipSymbolSearch(name) == false) {
+        BaseAST*    scopeToUse = use->getSearchScope();
+        const char* nameToUse  = name;
+
+        if (use->isARenamedSym(name) == true) {
+          nameToUse = use->getRenamedSym(name);
+        }
+
+        if (ResolveScope* next = getScopeFor(scopeToUse)) {
+          if (Symbol* sym = next->lookupNameLocallyForImport(nameToUse)) {
+            if (isRepeat(sym, symbols) == false) {
+              if (FnSymbol* fn = toFnSymbol(sym)) {
+                if (fn->isMethod() == false) {
+                  symbols.push_back(fn);
+                }
+
+              } else {
+                symbols.push_back(sym);
+              }
+            }
+          }
+        }
+      }
+    } else if (ImportStmt* import = toImportStmt(useImportList[i])) {
+      if (SymExpr* se = toSymExpr(import->src)) {
+        // The import statement has been resolved
+        if (import->isARename() == true) {
+          if (name == import->getRename()) {
+            symbols.push_back(se->symbol());
+          }
+        } else {
+          if (name == se->symbol()->name) {
+            symbols.push_back(se->symbol());
+          }
+        }
+      }
+
+      // Found a NULL sentinel.  Break if there are results.
+    } else {
+      if (symbols.size() > 0) {
+        break;
+      }
+    }
+  }
+  if (symbols.size() == 1) {
+    return symbols[0];
+  } else {
+    return NULL;
+  }
+}
+
+Symbol* ResolveScope::lookupForImport(Expr* expr) const {
+  Symbol* retval = NULL;
+
+  if (UnresolvedSymExpr* uSym = toUnresolvedSymExpr(expr)) {
+    const ResolveScope* ptr = NULL;
+    for (ptr = this; ptr != NULL && retval == NULL; ptr = ptr->mParent) {
+      // Check if the module is defined in this scope
+      Symbol* sym = ptr->lookupNameLocallyForImport(uSym->unresolved);
+      if (ModuleSymbol* mod = toModuleSymbol(sym)) {
+        if (mod->defPoint->parentSymbol != theProgram) {
+          // if we're not in the root module scope, this is an improper match
+          USR_FATAL(expr, "currently unable to import nested modules without"
+                    " specifying the full path to the module");
```

Comment:
Do you have a test with 2 modules with the same name, where one is toplevel and one is a submodule? There should not be an error, and the top level one should be chosen.

e.g. starting with this but filling it in to know which M you got

``` chapel
module M {
}

module User {
  module M {
  }

  import M;
}
```

Reply:
This does currently give an error.  I'm thinking I'll give a warning for the inner one and only fail if we can't find a top level module with the right name, does that sound good?

Reply:
Okay, so I did the warning thing, but it was still resolving to the inner module instead of the outer one for writing qualified access out:
```chapel
module M {
  var x: int = 4;
}

module User {
  module M {
    var x: int = 7;
  }

  import M;
  proc main() {
    writeln(M.x); // Was printing 7 instead of 4
  }
}
```
This is the same behavior as if the import statement wasn't there at all, though - basically, the inner M's definition is considered to be a closer scope than the import itself.  Changing this would be a breaking change.

But I think I get what you were trying to test, so I wrapped the writing call in another module and that chose the top level module
```chapel
module M {
  var x: int = 4;
}

module User {
  module M {
    var x: int = 7;
  }

  module Inner {
    import M;
    proc main() {
      writeln(M.x); // Wrote 4, yay!
    }
  }
}
```

Reply:
sounds good for now if you can add that as a test